### ISSUE TITLE
Lessons after running Kafka in production for two years

### ISSUE THREAD

**Issue by anitsh** [codeanit/todo#924]
## Context
- Self-hosted and in production for more than 2 years
- Multiple pipelines, topics >100 with over 5000 pods (clients) producing and consuming over 10 million messages daily.

Kafka is a message broker which helps us communicate between two pieces of software. It helps us publish messages on “topics” that we can listen to using consumers. 

## Wins
Kafka works great even with a huge volume of data especially if the packets are small in size.
Ensures “at least once” delivery of messages meaning no messages get lost in transit.
Once you get Kafka running properly with ample RAM and HDD it will run for a really long time and is quite low maintenance.

Can handle a huge number of connections, Example with LinkedIn reporting over 10k connections per node.

Even our paltry single node, self-hosted Kafka can handle itself well with 5k connections and 100s of topics with many 1000 partition topics. (Author)

## Pitfalls:
Big-sized messages: If your message size is over 1Mb, there are configuration changes needed and even with the correct configuration, Kafka struggling with sub 10Mb messages with throughput being quite low and requiring an unfeasible amount of consumers to operate without lag.

Choosing the number of partitions for a topic. This is the main challenge.

Log clearance might be slow.
Kafka has no way to re-distribute messages among partitions

If the number of partitions is too huge and the if the number of consumers is very less, it will cause issues. A huge number of partitions requires Kafka to create, open connections, and write to more log files. Open files limit needs to be appropriate.

TLDR: Have atleast as many paritions (P) as max number of consumers(C), but number of partitions should always be multiple of number of consumers. P = n * C, n =1,2,3…

Whenever a new consumer joins a consumer group and starts listening to a topic, Kafka assigns it a partition. If a consumer leaves a group, its assigned partition will be re-assigned. When frequently partitioning is done then it may crash.

## Be Careful
Open File Limit: Since Kafka is log-based message broker, it needs to write in log files and essentially open many connections in proportion to the number of topics you have and further how many partitions those topics have. 
https://www.tecmint.com/increase-set-open-file-limits-in-linux

Log Retention: If incorrectly set, hard disk might get exhausted leading to downtime. Estimate how much data you would send through Kafka in an hour/day/week and set it accordingly.

https://medium.com/@hardiktaneja_99752/lessons-after-running-kafka-in-production-626974ffd700
